source/_drafts/videojs-contrib-ads-6.md

```diff
@@ -0,0 +1,41 @@
+
+## How it works
+
+For a history of how the state machine has evolved over time, [there is detailed information in this Github issue](https://github.com/videojs/videojs-contrib-ads/issues/320). In this article we're going to focus on how it works in version 6.
+
+Here is a diagram of the current states:
+
+![](./ad-states.png)
+
+There are two types of states: content states (blue) and ad states (yellow). `player.ads.isInAdMode()` is implemented using this distinction. [Ad mode](https://github.com/videojs/videojs-contrib-ads#ad-mode-definition) is defined as _if content playback is blocked by the ad plugin_. This does not necessarily mean that an ad is playing. For example, if there is a preroll ad, after a play request we show a loading spinner until ad playback actually begins. This is considered part of ad mode.
+
+Let's walk through the states for a simple preroll scenario. The plugin begins in BeforePreroll state. This is considered a content state because content playback hasn't been requested yet and so ads are not blocking playback. However, the ad plugin can still asynchronously request ads from an ad server during this time. The ad plugin triggers the `adsready` event to say that it's ready. Contrib-ads knows that once the play button is pressed, ads are already prepared.
+
+Now, the user clicks play. At this point ad mode begins and contrib-ads moves forward to the Preroll state. The Preroll state shows a loading spinner while ads load. Once the ad break begins and an ad plays, the control bar turns yellow and will not allow seeking. If the ad takes too long to begin playing, a timeout will occur and content will resume without further adieu. When the ad is complete, contrib-ads restores the content video. Only when content playback results in a `playing` event does ad mode end.
+
+Now we're in the ContentPlayback state. At this point, the ad plugin could play a midroll ad, causing a brief foray into the Midroll state, or content could continue without interruption.
+
+When content ends, normally we would see an `ended` event right away. Instead, contrib-ads sends a `contentended` event, which is the ad plugin's chance to play postroll ads. Perhaps the ad plugin has decided not to play postroll ad, so it responds by triggering `nopreroll`. contrib-ads knows that now we're really done, so an `ended` event is triggered.
```

Comment:
so it responds by triggering ~~`nopreroll`~~ `nopostroll`

Reply:
Fixed